---
title: "Figure S2"
output: 
  html_document:
    df_print: paged
author: "John Favate"
date: "`r Sys.time()`"
---

<style type="text/css">
.main-container {
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r}
# Prevent printing of warnings and such in the HTML
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.align = "center")
```

```{r}
library(tidyverse)
library(patchwork)
library(broom)
library(ggrepel)
library(scales)
library(ggpubr)
library(grid)
library(gridExtra)
library(gtable)

all.data <- read_csv("../../data_frames/all_data.csv") %>%
  filter(gene_type == "ECB")%>%
  mutate(line = str_replace(line, "Ara-", "A-"),
         line = str_replace(line, "Ara\\+", "A+"))

# get the e coli proteins only
ecb.genes <- unique(all.data$target_id)

# Removes ERCC and tRNA genes from the data
kdf <- read_csv("../../data_frames/table_s1_read_counts.csv") %>%
  filter(target_id %in% ecb.genes)%>%
  mutate(line = str_replace(line, "Ara-", "A-"),
         line = str_replace(line, "Ara\\+", "A+"))

mut.levels <- all.data %>% 
  select(line, mutator) %>% 
  distinct() %>% 
  arrange(mutator, line) %>% 
  pull(line) %>% 
  str_replace(., "Ara-", "A-") %>% 
  str_replace("Ara\\+", "A+")

# deseq results
ddf <- read_csv("../../data_frames/table_s2_fold_changes.csv") %>%
  filter(seqtype == "rna") %>%
  mutate(line = str_replace(line, "Ara-", "A-"),
         line = str_replace(line, "Ara\\+", "A+"),
         line = factor(line, levels = mut.levels))
```

These are the mutator lines.
```{r}
mut.lines <- all.data %>%
  filter(mutator == TRUE) %>%
  pull(line) %>%
  unique()
```

Define the set of genes that were significant in any one line. This includes deletions as significant.
```{r}
sig.genes <- ddf %>%
  filter(padj <= .01 | deleted == TRUE) %>%
  pull(target_id) %>%
  unique()
```

### Panel A

Full fold change distribution
```{r}
fc.only <- all.data %>%
  filter(gene_type == "ECB") %>%
  select(line, ds_log2foldchange_rna) %>% 
  mutate(line = factor(line, mut.levels))

mut.colors.a <- c(rep("black", 6), rep("firebrick3", 5))

pa <- fc.only %>%
  ggplot(., aes(line, ds_log2foldchange_rna)) +
  geom_boxplot(outlier.size = .5) +
  theme_bw() +
  theme(
    text = element_text(size = 14),
    panel.grid = element_blank(),
    axis.text.x = element_text(
      angle = 45,
      vjust = 1.1,
      hjust = 1.1,
      color = mut.colors.a
    )
  ) +
  labs(x = NULL,
       y = expression(paste(log[2], "(fold-change)")),
       title = "A") +
  scale_y_continuous(breaks = seq(-15, 15, 5))

pa
```

### Panel B

Number of sig genes per line
```{r}
bdf <- all.data %>% 
  filter(ds_padj_rna <= .01) %>% 
  group_by(line) %>% 
  tally()

# order lines by mutator/n genes
b.order <- bdf %>% 
  mutate(mutator = line %in% mut.lines) %>% 
  arrange(mutator, n) %>% 
  pull(line)

# order it
bdf$line <- factor(bdf$line, b.order)

# determine axis text color
b.colors <- sapply(b.order, function(x){
  if (x %in% mut.lines){
    return("firebrick")
  } else {
    return("black")
  }
})

pb <- bdf %>% 
  ggplot(., aes(line, n))+
  geom_col()+
  theme_bw()+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),
        legend.position = c(.5, .95),
        legend.direction = "horizontal",
        legend.background = element_blank(),
        axis.text.x = element_text(color = b.colors, angle = 45, vjust = 1.1, hjust = 1.1))+
  labs(x = NULL,
       y = "Differentially expressed genes",
       title = "B")+
  scale_y_continuous(breaks = breaks_pretty(n = 6))

pb
```

### Figure C

To compare the observed distribution to an expected distribution, we run many iterations of sampling of the observed distribution after shuffling the transcripts about the fold changes and padj values. This takes really long, over an hour on 32 threads, it's in a separate Rscript which can be run in the background so it doesn't paralyze Rstudio. That script is at `code/analysis/randomizations.R`. It uses 32 threads, so if you can't use that many it'll take much longer. The results are loaded here. The number of occurrences for 
```{r}
count.cis <- read_csv("../../data_frames/count_cis.csv") %>% 
  mutate(type = "Randomized") %>% 
  select(direc, n, mean_prob, type) %>% 
  rename("prob" = "mean_prob")
```

Load the probabilities from figure 1
```{r}
f1.probs <- read_rds("../../data_frames/fig_1_probs.rds") %>% 
  select(n, starts_with("Exp")) %>% 
  pivot_longer(starts_with("Exp"), names_to = "direc", values_to = "prob") %>% 
  mutate(type = "SINIB",
         direc = str_remove(direc, "\\w+_"))
```

Combine them
```{r fig.width = 4, fig.height = 4}
pc.df1 <- bind_rows(f1.probs, count.cis) %>% 
  filter(n != 0) %>% 
  unite("ltype", type, direc, sep = " ") %>% 
  mutate(prob = ifelse(prob == 0, NA, prob))
```

KS test them
```{r}
ksdf <- pc.df1 %>% 
  pivot_wider(names_from = ltype, values_from = prob)

p.up <- signif(ks.test(ksdf$`SINIB up`, ksdf$`Randomized up`)$p.value, 3)

p.down <- signif(ks.test(ksdf$`SINIB down`, ksdf$`Randomized down`)$p.value, 3)
```


```{r fig.width = 4, fig.height = 4}
pc.top <- pc.df1 %>%
  ggplot(., aes(n, prob, color = ltype, linetype = ltype)) +
  geom_line() +
  scale_y_log10(labels = trans_format("log10", math_format(10 ^ .x))) +
  scale_x_continuous(breaks = 1:11, name = NULL) +
  scale_color_manual(values = rep(c("indianred", "steelblue"), 2), name = NULL) +
  scale_linetype_manual(values = c(1, 1, 2, 2), name = NULL) +
  theme_bw() +
  theme(
    text = element_text(size = 14),
    panel.grid = element_blank(),
    legend.position = c(.32, .26),
    legend.text = element_text(size = 9),
    legend.key = element_blank(),
    legend.background = element_blank(),
    legend.spacing.y = unit(.1, "mm"),
    legend.key.height = unit(3, "mm")
  ) +
  labs(x = "Number of lines",
       y = "Probability",
       title = "C") +
  guides(color = guide_legend(byrow = TRUE))+
  annotate("text", label = paste("p =", p.up), x = 10, y = .1, color = "steelblue")+
  annotate("text", label = paste("p =", p.down), x = 10, y = .001, color = "indianred")

pc.top
```

The lower part
```{r fig.width = 4, fig.height = 4}
pc.df2 <- bind_rows(f1.probs, count.cis) %>% 
  filter(n != 0) %>% 
  unite("ltype", type, direc, sep = " ") %>% 
  mutate(prob = ifelse(prob == 0, NA, prob),
         n_genes = prob * length(unique(ddf$target_id)),
         ltype = factor(ltype, c("Randomized down", "SINIB down", "Randomized up", "SINIB up"))) 

colorz <- c("#E31A1C", "#FB9A99", "#1F78B4", "#A6CEE3")

inset <- pc.df2 %>% 
  filter(n >= 3) %>% 
  ggplot(., aes(n, n_genes, fill = ltype))+
  geom_col(position = "dodge")+
  scale_x_continuous(breaks = 1:11)+
  scale_y_continuous(breaks = c(0, 10, 20))+
  scale_fill_manual(values = colorz, name = NULL, guide ="none")+
  theme_bw()+
  theme(text = element_text(size = 14),
        panel.grid = element_blank(),
        legend.background = element_blank(),
        legend.position = c(.7, .8),
        plot.background = element_blank(),
        axis.text.y = element_text(size = 9))+
  labs(x = NULL, y= NULL)

pc.big <- pc.df2 %>% 
  ggplot(., aes(n, n_genes, fill = ltype))+
  geom_col(position = "dodge")+
  scale_x_continuous(breaks = 1:11)+
  scale_fill_manual(values = colorz, name = NULL)+
  theme_bw()+
  theme(text = element_text(size = 14),
        panel.grid = element_blank(),
        legend.background = element_blank(),
        legend.position = c(.7, .25),
        legend.text = element_text(margin = margin(l = .001, unit = "mm"), size = 8.5),
        legend.key.size = unit(1.5, "mm"),
        legend.spacing.y = unit(1, 'mm'))+
  labs(x = "Number of lines",
       y = "Number of genes")

pc.bottom <- pc.big + inset_element(inset, left = .15, right = .95, top = .99, bottom = .38)

pc.bottom
```

```{r, fig.width = 11/3, fig.height = 4}
pc.full <- pc.top / pc.bottom

pc.full
```

### Panel D

A boxplot that shows the number, magnitude, and direction of fold changes per line.
```{r}
fc.count.df <- ddf %>% 
  filter(padj <= sig) %>% 
  mutate(direc = ifelse(log2FoldChange > 0, "Upregulated", "Downregulated")) %>% 
  group_by(line, direc) %>% 
  summarise(n = n()) %>% 
  group_by(line) %>% 
  mutate(prop = signif(n / sum(n), 2),
         total = sum(n))
```

A set of binomial tests to show that the number of up and down genes is different. 
```{r}
fc.count.df %>% 
  select(-prop) %>% 
  pivot_wider(names_from = direc, values_from = n) %>% 
  split(.$line) %>% 
  map(function(x){
    binom.test(x$Downregulated, x$total, alternative = "greater") %>% 
      tidy()
  }) %>% 
  bind_rows(.id = "line") %>% 
  mutate(sig = p.value <= .05) %>% 
  group_by(sig) %>% 
  tally()
```

stat_compare_means can't use a KS test, perform one manually and manually add the stars.
```{r}
pval.df <- ddf %>%
  filter(padj <= sig) %>%
  mutate(direc = ifelse(log2FoldChange > 0, "Upregulated", "Downregulated")) %>%
  select(line, target_id, log2FoldChange, direc) %>%
  pivot_wider(names_from = direc, values_from = log2FoldChange) %>%
  split(.$line) %>%
  map(function(x) {
    ks.test(na.omit(x$Downregulated), na.omit(x$Upregulated), exact = TRUE) %>%
      tidy()
  }) %>%
  bind_rows(.id = "line") %>%
  mutate(cuts = cut(
    p.value,
    c(1, .05, .01, .001, .0001,-Inf),
    labels = c("****", "***", "**", "*", "NS")
  ))

pval.df
```

A box plot showing the significant fold changes
```{r fig.width = 3.6, fig.height = 4}
mut.colors.d <- c(rep("black", 6), rep("firebrick3", 5)) 

pd <- ddf %>% 
  filter(padj <= sig) %>% 
  mutate(direc = ifelse(log2FoldChange > 0, "Upregulated", "Downregulated")) %>% 
  ggplot(., aes(line, abs(log2FoldChange), fill = direc))+
  geom_boxplot(outlier.size = .5)+
  geom_text(inherit.aes = FALSE, data = fc.count.df, aes(x = line, y = -.1, color = direc, label = n), position = position_dodge(width = .85), size = 3.4)+
  scale_fill_manual(name = NULL, values = c("indianred", "steelblue"))+
  scale_color_manual(name = NULL, values = c("indianred", "steelblue"), guide = "none")+
  theme_bw()+
  theme(text = element_text(size = 14),
        panel.grid = element_blank(),
        legend.position = c(.66, .94),
        axis.text.y = element_text(color = mut.colors.d),
        legend.background = element_blank(),
        legend.text = element_text(size = 10),
        legend.key.size = unit(4, "mm"))+
  labs(x = NULL,
       y = expression(paste("|", log[2], "(fold-change)|")),
       title = "D")+
  guides(fill = guide_legend(nrow = 1))+
  geom_text(data = pval.df, aes(x = line, y = 16, label = cuts), inherit.aes = FALSE)+
  guides(fill=guide_legend(ncol=1))+
  coord_flip()

pd
```

### Panel E

A PCA using all fold-changes, with deletions given a value of -10 for two reasons. One, the PCA needs complete data and deleted genes have fold-change of 0. Two, the deleted genes are treated as highly downregulated because they're either completely absent or pseudogenized and don't produce functional proteins.
```{r}
# genes deleted in all have to be removed, PCA can't scale a column with 0 variance
good.genes <- all.data %>% 
  filter(deleted == TRUE) %>% 
  group_by(target_id) %>% 
  tally() %>% 
  filter(n != 11) %>% 
  pull(target_id)

pca.input <- all.data %>%
  select(line, target_id, ds_log2foldchange_rna, deleted) %>%
  filter(target_id %in% good.genes) %>% 
  mutate(ds_log2foldchange_rna = ifelse(deleted == TRUE, -10, ds_log2foldchange_rna)) %>% 
  select(-deleted) %>% 
  pivot_wider(names_from = target_id, values_from = ds_log2foldchange_rna) %>% 
  column_to_rownames("line") %>% 
  .[ , colSums(is.na(.)) == 0] # some genes have NAs as fold-changes anyway, remove those
```

```{r}
pca.results <- prcomp(pca.input, scale = TRUE)
```

Determine the variance explained
```{r}
var.explained <- pca.results %>%
  tidy(matrix = "eigenvalues") 

pc1.explain <- paste0("PC1 (", 100*signif(var.explained[var.explained$PC == 1,]$percent, 4), "%)")

pc2.explain <- paste0("PC2 (", 100*signif(var.explained[var.explained$PC == 2,]$percent, 4), "%)")
```

```{r fig.width = 3.5, fig.height = 3.5}
pe <- pca.results %>% 
  augment(pca.input) %>% 
  rename("line" = ".rownames") %>% 
  mutate(mutator = line %in% mut.lines) %>% 
  ggplot(aes(.fittedPC1, .fittedPC2, label = line, color = mutator))+
  geom_point()+
  geom_text_repel(min.segment.length = .0001, max.overlaps = Inf, segment.alpha = .5)+
  theme_bw()+
  theme(text = element_text(size = 14),
        panel.grid = element_blank())+
  labs(x = pc1.explain,
       y = pc2.explain,
       title = "E")+
  scale_color_manual(values = c("black", "firebrick3"), guide = "none")+
  expand_limits(y = c(0, 15), x = c(-18, 0))

pe
```

How does -1:-10 affect the results?
```{r}
pca.p.list <- lapply(-10:-1, function(x) {
  pca.input <- all.data %>%
    select(line, target_id, ds_log2foldchange_rna, deleted) %>%
    filter(target_id %in% good.genes) %>%
    mutate(ds_log2foldchange_rna = ifelse(deleted == TRUE, x, ds_log2foldchange_rna)) %>%
    select(-deleted) %>%
    pivot_wider(names_from = target_id, values_from = ds_log2foldchange_rna) %>%
    column_to_rownames("line") %>%
    .[, colSums(is.na(.)) == 0] # some genes have NAs as fold-changes anyway, remove those
  
  pca.results <- prcomp(pca.input, scale = TRUE)
  
  var.explained <- pca.results %>%
    tidy(matrix = "eigenvalues")
  
  pc1.explain <-
    paste0("PC1 (", 100 * signif(var.explained[var.explained$PC == 1, ]$percent, 4), "%)")
  
  pc2.explain <-
    paste0("PC2 (", 100 * signif(var.explained[var.explained$PC == 2, ]$percent, 4), "%)")
  
  pca.results %>%
    augment(pca.input) %>%
    rename("line" = ".rownames") %>%
    mutate(mutator = line %in% mut.lines) %>%
    ggplot(aes(.fittedPC1, .fittedPC2, label = line, color = mutator)) +
    geom_point() +
    geom_text_repel(
      min.segment.length = .0001,
      max.overlaps = Inf,
      segment.alpha = .5
    ) +
    theme_bw() +
    theme(text = element_text(size = 14),
          panel.grid = element_blank()) +
    labs(x = pc1.explain,
         y = pc2.explain,
         title = x) +
    scale_color_manual(values = c("black", "firebrick3"), guide = "none") +
    expand_limits(y = c(0, 15), x = c(-18, 0))
})
```

Seems like it doesn't affect the distribution of the points too much.
```{r fig.width = 10, fig.height = 8}
wrap_plots(pca.p.list)
```

### Panel F

Check the contributors to PC1 and PC2
```{r fig.width = 4, fig.height = 4}
top.10.pc1 <- pca.results %>% 
  tidy(matrix = "rotation") %>% 
  pivot_wider(names_from = PC, values_from = value) %>% 
  arrange(desc(abs(`1`))) %>% 
  pull(column) %>% 
  .[1:10]

top.10.pc2 <- pca.results %>% 
  tidy(matrix = "rotation") %>% 
  pivot_wider(names_from = PC, values_from = value) %>% 
  arrange(desc(abs(`2`))) %>% 
  pull(column) %>% 
  .[1:10]
```

```{r}
threshold <- 3

scale.labels <- c(paste("\u2264", paste0("-", threshold), sep = ""),
                  "",
                  0,
                  "",
                  paste("\u2265", threshold, sep = ""))
```

```{r}
names.df <- all.data %>% 
  select(target_id, rel_name) %>% 
  distinct()

hm.df <- ddf %>%
  filter(target_id %in% c(top.10.pc1, top.10.pc2)) %>%
  mutate(
    pc = ifelse(target_id %in% top.10.pc1, "PC1", "PC2"),
    stars = ifelse(padj <= .01, "*", NA),
    line = factor(line, mut.levels),
    newfc = case_when(
      log2FoldChange < -3 ~ -3,
      log2FoldChange > 3 ~ 3,
      is.na(log2FoldChange) | TRUE ~ log2FoldChange
    )
  ) %>% 
  select(line, target_id, newfc, padj, pc, stars) %>% 
  left_join(names.df)
```

```{r fig.width = 4, fig.height = 4}
pf <- hm.df %>%
  split(.$pc) %>%
  map(function(x) {
    ggplot(x, aes(line, rel_name, fill = newfc, label = stars)) +
      geom_raster() +
      scale_fill_gradientn(
        colors = c("indianred", "white", "steelblue"),
        labels = scale.labels,
        breaks = c(-3, -1.5, 0, 1.5, 3),
        name = expression(paste(log[2], "(FC)")),
        guide = guide_colorbar(ticks.colour = "black"),
        limits = c(-3, 3)
      ) +
      geom_text() +
      facet_wrap( ~ pc, ncol = 1, scales = "free", strip.position = "right") +
      theme_bw() +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1, color = mut.colors.a),
        text = element_text(size = 13),
        axis.text.y = element_text(size = 8)
      ) + 
      labs(x = NULL, y = NULL)
  })
  
pf.top <- pf$PC1 + theme(axis.text.x = element_blank(),
                         axis.ticks.x = element_blank())+ labs(title = "F")

pf.bottom <- pf$PC2 + guides(fill = "none")

pf.full <- (pf.top / pf.bottom) + plot_layout(guides = "collect")

pf.full
```

### Panel G

The table underneath to show the gene names and what they do
```{r fig.width = 11, fig.height = 3}
str_clean_func <- function(x){
  x %>% 
    str_remove_all("\\&") %>% 
    str_remove_all("\\;") %>% 
    str_remove_all("\\<sup\\>") %>% 
    str_remove_all("\\<\\/sup\\>")
}

ttt <- ttheme_default(core=list(fg_params=list(hjust=0, x=.01, cex = .7)),
                      rowhead=list(fg_params=list(hjust=0, x=0)))

table.title <- textGrob("G",gp=gpar(fontsize=17))

t1 <- all.data %>%
  filter(target_id %in% c(top.10.pc1)) %>% 
  select(target_id, rel_name, rel_product) %>% 
  unique() %>% 
  mutate(target_id = factor(target_id, levels = c(top.10.pc1))) %>% 
  arrange(target_id) %>% 
  select(-target_id) %>% 
  rename("PC1 genes" = "rel_name", "Product" = "rel_product") %>% 
  mutate(Product = str_clean_func(Product)) %>% 
  tableGrob(theme = ttt, rows = NULL)

t1 <- gtable_add_rows(
  t1, heights = grobHeight(table.title) + unit(0.5,"line"), pos = 0
)

t1 <- gtable_add_grob(t1,
                      list(table.title),
                      t = 1,
                      l = 1)

blank <- textGrob(" ",gp=gpar(fontsize=17))

t2 <- all.data %>%
  filter(target_id %in% c(top.10.pc2)) %>% 
  select(target_id, rel_name, rel_product) %>% 
  unique() %>% 
  mutate(target_id = factor(target_id, levels = c(top.10.pc2))) %>% 
  arrange(target_id) %>% 
  select(-target_id) %>% 
  rename("PC2 genes" = "rel_name", "Product" = "rel_product") %>% 
  mutate(Product = str_clean_func(Product)) %>% 
  tableGrob(theme = ttt, rows = NULL)

t2 <- gtable_add_rows(
  t2, heights = grobHeight(blank) + unit(0.5,"line"), pos = 0
)

t2 <- gtable_add_grob(t2,
                      list(blank),
                      t = 1,
                      l = 1)
```

### Final figure 

```{r fig.width = 11, fig.height = 10.5}
top <- (pa | pb | pc.full)

middle <- (pd | pe | pf.full) + plot_layout(widths = c(.38, .32, .3))

bottom <- (plot_spacer() | t1 | t2) + plot_layout(widths = c(.01, .4, .6))

(ff <- (top / middle / bottom) + plot_layout(heights = c(.325, .325, .34)))
```

Save it
```{r}
write_rds(ff, "../../figures/fig_s2.rds")

ggsave(ff, filename = "../../figures/fig_s2.pdf", width = 11, height = 10.5)

ggsave(ff, filename = "../../figures/fig_s2.png", width = 11, height = 10.5)
```

```{r}
sessionInfo()
```